docs/proposals/advanced-nginx-extensions.md

```diff
@@ -15,14 +15,1156 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+    so those keepalive directives won't work.
+  - Session persistence ([`sticky`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#sticky)). Those are  valid use cases. But at the same time, they only apply to
+    NGINX Plus. Additionally, Gateway API started introducing session persistence.
+    See https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.BackendLBPolicy
+- stream module - there are not any features in https://nginx.org/en/docs/stream/ngx_stream_upstream_module.html
+
+Note: because NGF already inserts the `random` load-balancing method, an `upstream` snippet will not be able to
+configure a different method.
+
+If we choose to introduce upstream snippets, the SnippetsPolicy will need to support targeting a Service, because NGF
+generates one upstream per Service.
+
+> We don't support routes that correspond to the NGINX stream module -- TLSRoute, TCPRoute, and UDPRoute. However,
+> because we're going to support them in the future, this proposal for SnippetsPolicy includes `stream` and `server`
+> contexts of the stream module.
+
+SnippetsFilter supports `http`, `server`, and `location` contexts of the stream module.
+
+> TLSRoute, TCPRoute, and UPDRoute don't support filters. As a result, SnippetsFilter doesn't support stream-related
+> contexts.
+>
+> Snippets for location might share the same problem as mentioned in the issue https://github.com/nginxinc/nginx-gateway-fabric/issues/207,
+> depending on the NGINX directives being used in the snippets. This proposal doesn't address the problem but
+> anticipates the solution to https://github.com/nginxinc/nginx-gateway-fabric/issues/2079 will also solve the problem
+> for Snippets.
+
+#### Examples
+
+Below are a few examples of using snippets to bring unsupported NGINX configuration into NGF.
+
+##### Rate-limiting SnippetPolicy
+
+We use NGINX [limit req module](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html) to configure rate
+limiting.
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsPolicy
+metadata:
+  name: rate-limit
+spec:
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: cafe-route
+  snippets:
+  - context: http
+    value: limit_req_zone $binary_remote_addr zone=myzone:10m rate=1r/s;
+  - context: http.server.location
+    value: limit_req zone=myzone burst=3;
+```
+
+As a result, NGF will insert those snippets into the generated configuration for the referenced HTTPRoute:
+
+```text
+# this is http context
+
+limit_req_zone $binary_remote_addr zone=myzone:10m rate=1r/s;
+. . .
+
+server {
+  . . .
+
+  location /coffee {
+    . . .
+    limit_req zone=myzone burst=3;
+    . . .
+  }
+
+  location /tea {
+    . . .
+    limit_req zone=myzone burst=3;
+    . . .
+  }
+```
+
+##### Proxy Buffering SnippetPolicy
```

Comment:
Aren't we putting this field into a first-class policy? If so, maybe we don't need to have an example here?

Reply:
looks like it's scheduled for 2.1.0 https://github.com/nginxinc/nginx-gateway-fabric/issues/1398

I think all examples (except brotli) will end up as a first-class policy eventually -- https://github.com/nginxinc/nginx-gateway-fabric/blob/main/docs/proposals/nginx-extensions.md#prioritized-nginx-features

(although, it doesn't mean we don't need extensions - as the number of NGINX directives is still huge -- https://nginx.org/en/docs/dirindex.html)

since the GEP is not a feature documentation but rather a rationale/design for feature, I don't see much a problem of having buffering there.

happy to remove though!